src/iface/neighbor.rs

```diff
@@ -60,23 +65,49 @@ impl<'a> Cache<'a> {
     /// Neighbor entry lifetime, in milliseconds.
     pub(crate) const ENTRY_LIFETIME: Duration = Duration { millis: 60_000 };
 
+    /// Default number of entries in the cache before GC kicks in
+    pub(crate) const GC_THRESHOLD: usize = 1024;
+
     /// Create a cache. The backing storage is cleared upon creation.
     ///
     /// # Panics
     /// This function panics if `storage.len() == 0`.
     pub fn new<T>(storage: T) -> Cache<'a>
             where T: Into<ManagedMap<'a, IpAddress, Neighbor>> {
+
+        Cache::new_with_limit(storage, Cache::GC_THRESHOLD)
+    }
+
+    pub fn new_with_limit<T>(storage: T, gc_threshold: usize) -> Cache<'a>
+            where T: Into<ManagedMap<'a, IpAddress, Neighbor>> {
         let mut storage = storage.into();
         storage.clear();
 
-        Cache { storage, silent_until: Instant::from_millis(0) }
+        Cache { storage, gc_threshold, silent_until: Instant::from_millis(0) }
     }
 
     pub(crate) fn fill(&mut self, protocol_addr: IpAddress, hardware_addr: EthernetAddress,
                        timestamp: Instant) {
         debug_assert!(protocol_addr.is_unicast());
         debug_assert!(hardware_addr.is_unicast());
 
+        #[cfg(any(feature = "std", feature = "alloc"))]
+        let current_storage_size = self.storage.len();
+
+        match self.storage {
+            ManagedMap::Borrowed(_) =>  (),
+            #[cfg(any(feature = "std", feature = "alloc"))]
+            ManagedMap::Owned(ref mut map) => {
+                if current_storage_size >= self.gc_threshold {
+                    let new_btree_map = map.into_iter()
+                        .map(|(key, value)| (*key, *value))
+                        .filter(|(_, v)| timestamp < v.expires_at)
+                        .collect();
+
+                    replace(map, new_btree_map);
```

Comment:
1 last nit: `use core::mem` so that this becomes `mem::replace`, then it becomes more clear that `core::mem::replace` is being used here.